src/mscorlib/src/System/Lazy.cs

```diff
@@ -52,61 +198,22 @@ internal static class LazyHelpers
     [DebuggerDisplay("ThreadSafetyMode={Mode}, IsValueCreated={IsValueCreated}, IsValueFaulted={IsValueFaulted}, Value={ValueForDebugDisplay}")]
     public class Lazy<T>
     {
-
-#region Inner classes
-        /// <summary>
-        /// wrapper class to box the initialized value, this is mainly created to avoid boxing/unboxing the value each time the value is called in case T is 
-        /// a value type
-        /// </summary>
-        [Serializable]
-        class Boxed
-        {
-            internal Boxed(T value)
-            {
-                m_value = value;
-            }
-            internal T m_value;
-        }
-
-
-        /// <summary>
-        /// Wrapper class to wrap the excpetion thrown by the value factory
-        /// </summary>
-        class LazyInternalExceptionHolder
+        private static T CreateViaDefaultConstructor()
         {
-            internal ExceptionDispatchInfo m_edi;
-            internal LazyInternalExceptionHolder(Exception ex)
-            {
-                m_edi = ExceptionDispatchInfo.Capture(ex);
-            }
+            return (T)LazyNonGeneric.CreateViaDefaultConstructor(typeof(T));
         }
-#endregion
-
-        // A dummy delegate used as a  :
-        // 1- Flag to avoid recursive call to Value in None and ExecutionAndPublication modes in m_valueFactory
-        // 2- Flag to m_threadSafeObj if ExecutionAndPublication mode and the value is known to be initialized
-        static readonly Func<T> ALREADY_INVOKED_SENTINEL = delegate 
-        {
-            Debug.Assert(false, "ALREADY_INVOKED_SENTINEL should never be invoked.");
-            return default(T);
-        };
-
-        //null --> value is not created
-        //m_value is Boxed --> the value is created, and m_value holds the value
-        //m_value is LazyExceptionHolder --> it holds an exception
-        private object m_boxed;
 
-        // The factory delegate that returns the value.
-        // In None and ExecutionAndPublication modes, this will be set to ALREADY_INVOKED_SENTINEL as a flag to avoid recursive calls
+        // _state, a volatile reference, is set to null after m_value has been set
         [NonSerialized]
-        private Func<T> m_valueFactory;
+        private volatile LazyNonGeneric _state;
```

Comment:
@jkotas, is the JIT able to remove the volatility of writes to fields like this when performed in the ctor?  If not, might be better to use Volatile.Read/Write on call sites that need it rather than making the field volatile.